File: 2_Vanila Option/Option_MonteCarlo.py
# Monte Carlo
from math import log, exp, sqrt
from scipy.stats import norm
import numpy as np

def monte_carlo_European(S0, K, r, q, sigma, T, call_put, sims, rep):
    meanValueLst = []
    times = 0

    stockSamples = []
    while times < rep:
        for i in range(sims):
            # stockSamples.append(S0 * exp((r - q - 0.5 * sigma**2) * T + sigma * sqrt(T) * np.random.standard_normal()))
            # lnSample = log(S0) + (r - q - 0.5*sigma**2) * T + sigma * sqrt(T) * np.random.standard_normal()
            lnSample = np.random.normal(loc = log(S0) + (r - q - 0.5*(sigma**2))* T, scale = sigma * sqrt(T))
            sample = exp(lnSample)
            stockSamples.append(sample)
        
        optionValue = []
        if call_put == "call":
            for price in stockSamples:
                if K <= price:
                    optionValue.append(price-K)
                else:
                    optionValue.append(0)

            meanValue = np.mean(optionValue)
            discounted = meanValue * exp(-r * T)
            meanValueLst.append(discounted)
            times += 1

        else:
            for price in stockSamples:
                if K >= price:
                    optionValue.append(K-price)
                else:
                    optionValue.append(0)

            meanValue = np.mean(optionValue)
            discounted = meanValue * exp(-r * T)
            meanValueLst.append(discounted)
            times += 1

    sdOfRep = np.std(meanValueLst)
    meanOfRep = np.mean(meanValueLst)
    upperBound = meanOfRep + 2*sdOfRep
    upperBound = round(upperBound, 6)
    lowerBound = meanOfRep - 2*sdOfRep
    lowerBound = round(lowerBound, 6)
    bounds = [lowerBound, upperBound]
    print("==================================================")
    print(f"European {call_put}")
    print("--------------------------------------------------")
    print(f"平均 : {round(meanOfRep, 6)}")
    print(f"標準誤 : {round(sdOfRep, 6)}")
    print(f"九十五趴信賴區間 : {bounds}")
    return round(meanOfRep, 6)

S0 = 115
K = 120
r = 0.01
q = 0.02
sigma = 0.5
T = 1
sims = 10000
rep = 20
monte_carlo_European(S0, K, r, q, sigma, T, "call", sims, rep)
monte_carlo_European(S0, K, r, q, sigma, T, "put", sims, rep)

File: 2_Vanila Option/Option_MonteCarlo.cpp
# include <iostream>
# include <cmath>
# include <vector>
# include <random>

using namespace std;

vector<double> monte_carlo_European(double S0, double K, double r, double q, double sigma, double T, string call_put, int sims, int rep){
    vector<double> meanLst;
    int times = 0;

    default_random_engine generator;
    normal_distribution<double> distribution(log(S0) + (r - q - 0.5*(pow(sigma, 2)))* T, sigma * sqrt(T));

    while(times < rep){
        vector<double> stockSamples;
        for(int i = 0; i < sims; i++){
            double lnSample = distribution(generator);
            // cout << lnSample << endl;
            double sample = exp(lnSample);
            // cout << sample << endl;
            stockSamples.push_back(sample);
        }

        vector<double> optionValue;
        if(call_put == "call"){
            for(int j = 0; j < stockSamples.size(); j++){
                optionValue.push_back(max(stockSamples[j] - K, 0.0));
            }
            double sum = accumulate(begin(optionValue), end(optionValue), 0.0);
            double mean = sum/sims;
            double discounted = mean * exp(-r*T);
            meanLst.push_back(discounted);
            times += 1;
        }
        else{
            for(int l = 0; l < stockSamples.size(); l++){
                optionValue.push_back(max(K - stockSamples[l], 0.0));
            }
            double sum = accumulate(begin(optionValue), end(optionValue), 0.0);
            double mean = sum/sims;
            double discounted = mean * exp(-r*T);
            meanLst.push_back(discounted);
            times += 1;
        }
    }

    double sum_mean = accumulate(begin(meanLst), end(meanLst), 0.0);
    double meanOfRep = sum_mean/20;
    
    double var = 0.0;
    for(int n = 0; n < 20; n++){
        var += (meanLst[n] - meanOfRep) * (meanLst[n] - meanOfRep);
    }
    var = var/20;
    double sdOfRep = sqrt(var);
    double upper = meanOfRep + sdOfRep;
    double lower = meanOfRep - sdOfRep;
    vector<double> results = {meanOfRep, lower, upper};
    cout << "==================================================" << endl;
    cout << "European " << call_put << endl;
    cout << "--------------------------------------------------" << endl;
    cout << "mean : " << meanOfRep << endl;
    cout << "standard error : " << sdOfRep << endl;
    cout << "0.95 confidence interval : [ " << lower << ", " << upper << " ]" << endl;
    cout << endl;
    return results;
}




int main(){
    double S0 = 115;
    double K = 120;
    double r = 0.01;
    double q = 0.02;
    double sigma = 0.5;
    double T = 1;
    int sims = 10000;
    int rep = 20;

    monte_carlo_European(S0, K, r, q, sigma, T, "call", sims, rep);
    monte_carlo_European(S0, K, r, q, sigma, T, "put", sims, rep);

    return 0;
}